### PR TITLE
Support localized fields in API responses

### LINKED ISSUE #6422: Support localized fields in API responses
In order to fully switch to API v5, we need to support the new localized fields. Instead of directly getting a value for any localized field, we will always receive an object with at least one key/value (the key is the locale, the value is the localized content).

The API `v5` should be used to see those localized fields.

See also: https://github.com/mozilla/addons-server/issues/8794

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #6422

### CODE REVIEW

File: src/core/types/addons.js [hunk 1]
```diff
@@ -54,6 +55,24 @@ type EventDataType = {|
...
+|};
+
+export type PreviewType = {|
+  h: number,
+  src: string,
+  thumbnail_h: number,
+  thumbnail_src: string,
+  thumbnail_w: number,
+  title: string,
+  w: number,
```
Comment (willdurand) [inline src/core/types/addons.js:73]: If this is our internal type, can we have a better prop name than a single char?

File: src/core/types/addons.js [hunk 2]
```diff
@@ -54,6 +55,24 @@ type EventDataType = {|
...
+export type ExternalPreviewType = {|
+  caption: LocalizedString | null,
+  image_size: [number, number],
+  image_url: string,
+  thumbnail_size: [number, number],
+  thumbnail_url: string,
+|};
+
+export type PreviewType = {|
+  h: number,
```
Comment (willdurand) [inline src/core/types/addons.js:67]: If this is our internal type, can we have a better prop name than a single char?

File: src/core/types/addons.js [hunk 3]
```diff
@@ -122,6 +141,15 @@ export type PartialExternalAddonType = {|
  */
 export type AddonType = {|
   ...ExternalAddonType,
+  // normalized l10n fields
```
Comment (willdurand) [inline src/core/types/addons.js:144]: This is great but not true in the current patch. The assumption was that `selectLocalizedContent()` return type would be `string | null`. I think it's what we want and I explained why in a previous comment (so let's keep these changes here and update the selector a bit)

File: tests/unit/amo/components/TestAddAddonToCollection.js [hunk 4]
```diff
@@ -38,7 +38,7 @@ describe(__filename, () => {
 
   const getProps = (customProps = {}) => {
     return {
-      addon: createInternalAddon(fakeAddon),
+      addon: createInternalAddonWithLang(fakeAddon),
```
Comment (willdurand) [inline tests/unit/amo/components/TestAddAddonToCollection.js:41]: Awesome. I was thinking about doing something like that too.

Nit: I'd have used a slightly different name like `createInternalAddonWithDefaultLang` because, to me, `WithLang` means we'd pass a lang to the function.

File: tests/unit/amo/components/TestAddonMoreInfo.js [hunk 5]
```diff
@@ -139,17 +140,17 @@ describe(__filename, () => {
...
 
     expect(root.find('.AddonMoreInfo-support-link')).toHaveLength(0);
   });
 
   it('renders the support link of an add-on', () => {
-    const addon = createInternalAddon({
+    const addon = createInternalAddonWithLang({
       ...fakeAddon,
-      support_url: 'http://support.hampsterdance.com/',
+      support_url: createLocalizedString('http://support.hampsterdance.com/'),
```
Comment (willdurand) [inline tests/unit/amo/components/TestAddonMoreInfo.js:153]: I haven't seen this function yet. I suppose it uses a default lang? How do we sync `createInternalAddonWithLang` and `createLocalizedString`? Maybe we could name these functions `createInternalAddonWithDefaultLang` and `createLocalizedStringWithDefaultLang`?

It's verbose but explicit. Other ideas welcomed, though.

File: tests/unit/amo/components/TestCollectionDetails.js [hunk 6]
```diff
@@ -10,11 +10,11 @@ import {
   collectionEditUrl,
   collectionUrl,
   convertFiltersToQueryParams,
-  createInternalCollection,
 } from 'amo/reducers/collections';
 import {
   createFakeCollectionDetail,
   createFakeEvent,
+  createInternalCollectionWithLang,
```
Comment (willdurand) [inline tests/unit/amo/components/TestCollectionDetails.js:17]: `WithDefaultLang`? 😋

File: tests/unit/amo/components/TestInstallButtonWrapper.js [hunk 7]
```diff
@@ -79,14 +80,14 @@ describe(__filename, () => {
...
+      addon: createInternalAddonWithLang(addon),
       store,
     });
 
     sinon.assert.calledWith(_getClientCompatibility, {
-      addon: createInternalAddon(addon),
+      addon: createInternalAddonWithLang(addon),
       clientApp,
-      currentVersion: createInternalVersion(addon.current_version),
+      currentVersion: createInternalVersionWithLang(addon.current_version),
```
Comment (willdurand) [inline tests/unit/amo/components/TestInstallButtonWrapper.js:90]: `WithDefaultLang`?

File: tests/unit/amo/components/TestSponsoredAddonsShelf.js [hunk 8]
```diff
@@ -64,6 +65,7 @@ describe(__filename, () => {
   };
 
   const _loadPromotedExtensions = ({ addons = [] }) => {
+    store.dispatch(setLang('en-US'));
```
Comment (willdurand) [inline tests/unit/amo/components/TestSponsoredAddonsShelf.js:68]: `dispatchClientMetadata()` (which is called in the test setup method) already dispatches this action. Why is this needed here?

File: tests/unit/amo/pages/TestHome.js [hunk 9]
```diff
@@ -649,14 +649,14 @@ describe(__filename, () => {
       expect(heroRecommendation).toHaveProp('errorHandler', errorHandler);
       expect(heroRecommendation).toHaveProp(
         'shelfData',
-        createInternalHeroShelves(heroShelves).primary,
+        createInternalHeroShelves(heroShelves, 'en-US').primary,
```
Comment (willdurand) [inline tests/unit/amo/pages/TestHome.js:652]: Should we have a `createInternalHeroShelvesWithDefaultLang` helper?

File: tests/unit/amo/pages/TestLandingPage.js [hunk 10]
```diff
@@ -53,6 +55,7 @@ describe(__filename, () => {
     errorHandler = createStubErrorHandler(),
     ...otherParams
   } = {}) => {
+    store.dispatch(setLang('en-US'));
```
Comment (willdurand) [inline tests/unit/amo/pages/TestLandingPage.js:58]: Why is this needed?

File: tests/unit/helpers.js [hunk 11]
```diff
@@ -1354,3 +1369,23 @@ export const fakeSponsoredShelf = Object.freeze({
   impression_data: 'some data',
   impression_url: 'https://mozilla.org/',
 });
+
+export const createInternalAddonWithLang = (addon, lang = 'en-US') => {
```
Comment (willdurand) [inline tests/unit/helpers.js:1373]: Oh,I see why it is not suffixed with `WithDefaultLang`. I guess that's fine then?

File: tests/unit/helpers.js [hunk 12]
```diff
@@ -37,23 +38,29 @@ import {
...
 import { loadCurrentUserAccount } from 'amo/reducers/users';
+import { createInternalVersion } from 'core/reducers/versions';
 import { createUIStateMapper, mergeUIStateProps } from 'core/withUIState';
 import { addQueryParamsToHistory } from 'core/utils';
 
 export const sampleUserAgent =
   'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:40.0) Gecko/20100101 Firefox/40.1';
 export const sampleUserAgentParsed = UAParser(sampleUserAgent);
 
+export const createLocalizedString = (string, lang = 'en-US') => {
```
Comment (willdurand) [inline tests/unit/helpers.js:57]: Instead of having `'en-US'` defined in different functions, maybe we could introduce a constant (`DEFAULT_LANG` or `DEFAULT_LANG_IN_TESTS`) in this file? We could use `'en-US'` or `config.get('defaultLang')` as a value.

**Review by willdurand — COMMENTED** [review]
Awesome, thanks!

I have some comments but this is looking very good.

**Comment by bobsilverberg** [comment]
> Awesome, thanks!
> 
> I have some comments but this is looking very good.

Thanks for the initial look @willdurand. I'll respond to your comments. There is still work to be done to add coverage for a bunch of things. What I've done so far is to just change all the code and get all the tests passing, but there is definitely missing coverage in the current patch.

File: src/amo/reducers/addonsByAuthors.js [hunk 13]
```diff
@@ -21,6 +23,7 @@ export type AddonsByAuthorsState = {|
 |};
 
 export const initialState: AddonsByAuthorsState = {
+  lang: '',
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/reducers/addonsByAuthors.js:26]: Done in the next commit.

File: src/amo/components/ScreenShots/index.js [hunk 14]
```diff
@@ -59,36 +60,7 @@ export const PHOTO_SWIPE_OPTIONS = {
...
-    src: preview.image_url,
-    thumbnail_h: preview.thumbnail_size[1],
-    thumbnail_src: preview.thumbnail_url,
-    thumbnail_w: preview.thumbnail_size[0],
-    title: preview.caption,
-    w: preview.image_size[0],
-  }));
-
-export const thumbnailContent = (item: Preview): React.Node => (
+export const thumbnailContent = (item: PreviewType): React.Node => (
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/components/ScreenShots/index.js:63]: Yes and no. `caption` is a localized field, so yes I do need to touch this code. While I was editing it I realized that it's not optimal and not the way we do it anywhere else, so I decided to fix that with a refactor. Considering how big this patch is already, and the fact that pretty much everything in the site will need to be tested, I'm not sure there's a need to extract these changes out into a separate patch. If you feel really strongly about it I can do it, but I feel like this patch is a good opportunity to clean some things up (like this) as long as we're touching the files.

File: src/core/imageUtils.js [hunk 15]
```diff
@@ -25,8 +25,8 @@ export const getPreviewImage = (
     // 720 is now the standard width for previews.
     const width = 720;
     imageIndex =
-      // The preview.image_size[0] is the image width.
-      addon.previews.findIndex((preview) => preview.image_size[0] === width);
+      // The preview.w is the image width.
+      addon.previews.findIndex((preview) => preview.w === width);
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/core/imageUtils.js:29]: Yes, as mentioned above, `caption` is a localized string.

File: src/core/reducers/utils.js [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+import type { LocalizedString } from 'core/types/api';
+
+export const selectLocalizedContent = (
+  field: LocalizedString,
+  lang: string,
+) => {
+  invariant(lang, 'lang must not be empty');
+  if (!field) {
+    return undefined;
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/core/reducers/utils.js:11]: I did this to deal with `notes` in CollectionAddon not allowing `null`, but going through the types again I think it actually should allow `null`, so I think I can change this back.

File: src/core/types/addons.js (hunk 1 above)
Comment (bobsilverberg) [reply src/core/types/addons.js:73]: Yes, but before I change it I want to confirm whether you are okay with keeping these changes in this patch, or if we should leave them to a follow-up patch.

File: src/core/types/addons.js (hunk 2 above)
Comment (bobsilverberg) [reply src/core/types/addons.js:67]: Yes, but before I change it I want to confirm whether you are okay with keeping these changes in this patch, or if we should leave them to a follow-up patch.

File: tests/unit/amo/components/TestAddAddonToCollection.js (hunk 4 above)
Comment (bobsilverberg) [reply tests/unit/amo/components/TestAddAddonToCollection.js:41]: As you saw later, this function does accept a `lang`, it's just that most of the time we don't need to specify one so we don't pass it and it uses a default. Can I mark this comment as resolved?

File: tests/unit/amo/components/TestAddonMoreInfo.js (hunk 5 above)
Comment (bobsilverberg) [reply tests/unit/amo/components/TestAddonMoreInfo.js:153]: As discussed above (and I think as you noticed later), `createLocalizedString` does optionally accept a `lang` if we need to synchronize. Most of the time we don't, which makes the test code and its usage shorter. To me it's fine for a test to use both `createInternalAddonWithLang` and `createLocalizedString` without passing a `lang` to either, and assume that they are synchronized behind the scenes (which they are). Do you agree?

File: tests/unit/amo/components/TestCollectionDetails.js (hunk 6 above)
Comment (bobsilverberg) [reply tests/unit/amo/components/TestCollectionDetails.js:17]: See above.

File: tests/unit/amo/components/TestInstallButtonWrapper.js (hunk 7 above)
Comment (bobsilverberg) [reply tests/unit/amo/components/TestInstallButtonWrapper.js:90]: See above.

File: tests/unit/amo/components/TestSponsoredAddonsShelf.js (hunk 8 above)
Comment (bobsilverberg) [reply tests/unit/amo/components/TestSponsoredAddonsShelf.js:68]: You're right. I added that prematurely. I'll remove it.

File: tests/unit/amo/pages/TestHome.js (hunk 9 above)
Comment (bobsilverberg) [reply tests/unit/amo/pages/TestHome.js:652]: Yes, I'll add one.

File: src/amo/components/ScreenShots/index.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/amo/components/ScreenShots/index.js:63]: Although I think it'd be better to have separate patches, let's keep everything as it is currently (i.e. no other patch, we keep all changes/cleanups here)

File: tests/unit/amo/components/TestAddAddonToCollection.js (hunk 4 above)
Comment (willdurand) [reply tests/unit/amo/components/TestAddAddonToCollection.js:41]: > Can I mark this comment as resolved?

yes

File: tests/unit/amo/components/TestAddonMoreInfo.js (hunk 5 above)
Comment (willdurand) [reply tests/unit/amo/components/TestAddonMoreInfo.js:153]: I agree.

File: tests/unit/amo/pages/TestLandingPage.js (hunk 10 above)
Comment (bobsilverberg) [reply tests/unit/amo/pages/TestLandingPage.js:58]: Hmm, I guess this is another case of me adding this prematurely. I thought I needed it to get tests to pass, but apparently not. I'll remove it.

**Comment by willdurand** [comment]
> Thanks for the initial look @willdurand. I'll respond to your comments. There is still work to be done to add coverage for a bunch of things. What I've done so far is to just change all the code and get all the tests passing, but there is definitely missing coverage in the current patch.

Yep, it's going pretty well so far. Adding new test coverage is easy compared to refactoring an exiting test suite :) Good job!

File: tests/unit/helpers.js (hunk 12 above)
Comment (bobsilverberg) [reply tests/unit/helpers.js:57]: Yeah, I like this idea. I'll change it.

File: src/core/types/addons.js (hunk 1 above)
Comment (bobsilverberg) [reply src/core/types/addons.js:73]: Ah, I believe we're using `w` and `h` here because those are the property names that the `PhotoSwipeGallery` component expects, so we either need to keep them that way, or introduce another transformer to convert from `PreviewType` to an object that works with `PhotoSwipeGallery`. The latter seems totally unneccesary. How about I add a comment to the type to explain that the prop names are those needed for the `PhotoSwipeGallery` component?

Comment (willdurand) [reply src/core/types/addons.js:73]: You're right. I forgot about that (although I thought about it after having left the comment). Yes, a comment sounds good, thanks

**Comment by bobsilverberg** [comment]
> Yep, it's going pretty well so far. Adding new test coverage is easy compared to refactoring an exiting test suite :) Good job!

Thanks @willdurand! I have addressed all of your issues, including adding a comment for https://github.com/mozilla/addons-frontend/pull/9909#discussion_r539126081 (which I still need you to confirm is okay). I'll move forward with adding more coverage.

**Comment by bobsilverberg** [comment]
This is ready for a final review.

**Review by willdurand — COMMENTED** [review]
The language tools page does not work for me.

**Comment by bobsilverberg** [comment]
@willdurand I fixed the problem with Language Tools, and also a problem I discovered with Search box autocomplete. I believe I have tested everything locally now, so this is good for another look.

**Review by willdurand — APPROVED** [review]
looks good, thanks

**Comment by bobsilverberg** [comment]
Thanks @willdurand. I guess we land it on dev and see how it goes there.
